File: backend/app/ai/usage_pricing.py
"""Tarification IA — coût fournisseur vs prix retail (marge), input/output séparés."""
from __future__ import annotations

import json
import logging
import os
from dataclasses import dataclass
from typing import Dict, Tuple

logger = logging.getLogger(__name__)

# Coût fournisseur USD / 1M tokens (sources officielles, juin 2026).
# OpenAI: https://openai.com/api/pricing/
# xAI: https://docs.x.ai/developers/pricing
# Anthropic: https://platform.claude.com/docs/en/about-claude/pricing
_DEFAULT_PROVIDER_USD_PER_1M: Dict[str, Tuple[float, float]] = {
    # (input, output)
    "gpt-4.1-nano": (0.10, 0.40),
    "gpt-4o-mini": (0.15, 0.60),
    # grok-3-mini / grok-4.1 → alias xAI vers grok-4.3 (facturé au même tarif)
    "grok-3-mini": (1.25, 2.50),
    "grok-4.1": (1.25, 2.50),
    "grok-4.3": (1.25, 2.50),
    # grok-4.20-0309-reasoning (pipeline CAO)
    "grok-cad-reasoning": (1.25, 2.50),
    "claude-opus-4-7": (5.00, 25.00),
    "claude-opus-4-8": (5.00, 25.00),
}

# Alias API / UI → clé tarifaire canonique.
_MODEL_PRICING_ALIASES: Dict[str, str] = {
    "auto": "gpt-4o-mini",
    "grok-mini": "grok-3-mini",
    "grok-3-mini": "grok-3-mini",
    "grok": "grok-4.1",
    "grok-4.1": "grok-4.1",
    "grok-4.3": "grok-4.3",
    "grok-4.20": "grok-cad-reasoning",
    "grok-4.20-reasoning": "grok-cad-reasoning",
    "grok-4.20-0309-reasoning": "grok-cad-reasoning",
    "grok-build-0.1": "grok-cad-reasoning",
    "gpt-4-1-nano": "gpt-4.1-nano",
    "gpt-4.1-nano": "gpt-4.1-nano",
    "gpt-4o": "gpt-4o-mini",
    "gpt-4o-mini": "gpt-4o-mini",
    "claude-opus-4-20250514": "claude-opus-4-7",
    "claude-opus-4-7": "claude-opus-4-7",
    "claude-opus-4-8": "claude-opus-4-8",
    "opus-4.7": "claude-opus-4-7",
    "opus-4.8": "claude-opus-4-8",
}


def usage_markup_multiplier() -> float:
    raw = os.getenv("FORMA_USAGE_MARKUP", "1.25").strip()
    try:
        value = float(raw)
    except ValueError:
        value = 1.25
    return max(value, 1.0)


def on_demand_usage_markup_multiplier() -> float:
    """Marge retail au-delà du forfait Pro (usage à la demande)."""
    raw = os.getenv("FORMA_ON_DEMAND_USAGE_MARKUP", "1.65").strip()
    try:
        value = float(raw)
    except ValueError:
        value = 1.65
    return max(value, 1.0)


def pro_usage_allowance_usd() -> float:
    raw = os.getenv("FORMA_PRO_USAGE_ALLOWANCE_USD", "30").strip()
    try:
        value = float(raw)
    except ValueError:
        value = 30.0
    return max(value, 0.0)


def enterprise_usage_allowance_per_seat_usd() -> float:
    raw = os.getenv("FORMA_ENTERPRISE_USAGE_ALLOWANCE_USD_PER_SEAT", "25").strip()
    try:
        value = float(raw)
    except ValueError:
        value = 25.0
    return max(value, 0.0)


def enterprise_usage_allowance_usd(seat_count: int) -> float:
    seats = max(int(seat_count or 0), 1)
    return enterprise_usage_allowance_per_seat_usd() * seats


def _parse_rate_entry(value: object) -> Tuple[float, float] | None:
    if isinstance(value, (int, float)):
        blended_per_10k = float(value)
        per_1m = blended_per_10k * 100.0
        return per_1m, per_1m
    if isinstance(value, dict):
        inp = value.get("inputUsdPer1M", value.get("input"))
        out = value.get("outputUsdPer1M", value.get("output"))
        if inp is None or out is None:
            return None
        return float(inp), float(out)
    return None


def _provider_table() -> Dict[str, Tuple[float, float]]:
    raw = os.getenv("FORMA_MODEL_PROVIDER_RATES", "").strip()
    if raw:
        try:
            parsed = json.loads(raw)
            if isinstance(parsed, dict):
                table: Dict[str, Tuple[float, float]] = {}
                for key, value in parsed.items():
                    rates = _parse_rate_entry(value)
                    if rates:
                        table[str(key).strip().lower()] = rates
                if table:
                    return table
        except (json.JSONDecodeError, TypeError, ValueError):
            logger.warning("Invalid FORMA_MODEL_PROVIDER_RATES JSON")

    legacy = os.getenv("FORMA_MODEL_PROVIDER_USD_PER_10K", "").strip()
    if legacy:
        try:
            parsed = json.loads(legacy)
            if isinstance(parsed, dict):
                table = {}
                for key, value in parsed.items():
                    rates = _parse_rate_entry(value)
                    if rates:
                        table[str(key).strip().lower()] = rates
                if table:
                    logger.warning(
                        "FORMA_MODEL_PROVIDER_USD_PER_10K is deprecated; "
                        "use FORMA_MODEL_PROVIDER_RATES with input/output per 1M tokens."
                    )
                    return table
        except (json.JSONDecodeError, TypeError, ValueError):
            pass

    return dict(_DEFAULT_PROVIDER_USD_PER_1M)


def normalize_pricing_model(model_id: str) -> str:
    mid = (model_id or "").strip().lower()
    if not mid:
        return "gpt-4o-mini"
    if mid in _MODEL_PRICING_ALIASES:
        return _MODEL_PRICING_ALIASES[mid]
    for alias, key in _MODEL_PRICING_ALIASES.items():
        if alias in mid or mid in alias:
            if len(alias) >= 4:
                return key
    if "nano" in mid and "gpt" in mid:
        return "gpt-4.1-nano"
    if "gpt-4o-mini" in mid or mid.startswith("gpt-4o"):
        return "gpt-4o-mini"
    if "grok-3-mini" in mid or mid == "grok-mini":
        return "grok-3-mini"
    if "build" in mid or "4.20" in mid or "reasoning" in mid:
        return "grok-cad-reasoning"
    if "4.1" in mid and "grok" in mid:
        return "grok-4.1"
    if "grok" in mid:
        return "grok-4.3"
    if "opus" in mid and ("4-8" in mid or "48" in mid):
        return "claude-opus-4-8"
    if "claude" in mid or "opus" in mid:
        return "claude-opus-4-7"
    return mid


def pricing_model_label(model_key: str) -> str:
    labels = {
        "gpt-4.1-nano": "GPT 4.1 nano",
        "gpt-4o-mini": "GPT 4o mini",
        "grok-3-mini": "Grok 3 Mini",
        "grok-4.1": "Grok 4.1",
        "grok-4.3": "Grok 4.3",
        "grok-cad-reasoning": "Grok reasoning (CAO)",
        "claude-opus-4-7": "Claude Opus 4.7",
        "claude-opus-4-8": "Claude Opus 4.8",
    }
    return labels.get(model_key, model_key)


@dataclass(frozen=True)
class ModelUsageRate:
    model_key: str
    provider_input_usd_per_1m: float
    provider_output_usd_per_1m: float
    retail_input_usd_per_1m: float
    retail_output_usd_per_1m: float


def model_usage_rate(model_id: str) -> ModelUsageRate:
    key = normalize_pricing_model(model_id)
    inp, out = _provider_table().get(key, (1.0, 1.0))
    markup = usage_markup_multiplier()
    return ModelUsageRate(
        model_key=key,
        provider_input_usd_per_1m=inp,
        provider_output_usd_per_1m=out,
        retail_input_usd_per_1m=inp * markup,
        retail_output_usd_per_1m=out * markup,
    )


def usage_cost_usd(
    model_id: str,
    input_tokens: int,
    output_tokens: int,
    *,
    retail: bool = True,
    markup_multiplier: float | None = None,
) -> float:
    inp = max(0, int(input_tokens))
    out = max(0, int(output_tokens))
    if inp <= 0 and out <= 0:
        return 0.0
    rate = model_usage_rate(model_id)
    if retail:
        multiplier = (
            float(markup_multiplier)
            if markup_multiplier is not None
            else usage_markup_multiplier()
        )
    else:
        multiplier = 1.0
    provider_cost = (inp / 1_000_000.0) * rate.provider_input_usd_per_1m
    provider_cost += (out / 1_000_000.0) * rate.provider_output_usd_per_1m
    return provider_cost * multiplier


def split_retail_charge(
    provider_cost: float,
    *,
    current_included_usd: float,
    allowance_usd: float,
    on_demand_enabled: bool,
) -> tuple[float, float]:
    """Répartit le coût fournisseur entre forfait (× marge incluse) et on-demand (×1.65)."""
    if provider_cost <= 0:
        return 0.0, 0.0
    included_markup = usage_markup_multiplier()
    on_demand_markup = on_demand_usage_markup_multiplier()
    remaining = max(0.0, float(allowance_usd) - float(current_included_usd))
    full_at_included = provider_cost * included_markup
    if full_at_included <= remaining + 1e-9:
        return full_at_included, 0.0
    if not on_demand_enabled:
        return min(full_at_included, remaining), 0.0
    if remaining <= 1e-9:
        return 0.0, provider_cost * on_demand_markup
    included_retail = remaining
    included_provider = included_retail / included_markup
    on_demand_provider = max(0.0, provider_cost - included_provider)
    on_demand_retail = on_demand_provider * on_demand_markup
    return included_retail, on_demand_retail


def list_public_model_rates() -> list[dict[str, float | str]]:
    markup = usage_markup_multiplier()
    on_demand_markup = on_demand_usage_markup_multiplier()
    rows: list[dict[str, float | str]] = []
    for key, (provider_in, provider_out) in sorted(_provider_table().items()):
        rows.append(
            {
                "modelKey": key,
                "label": pricing_model_label(key),
                "providerInputUsdPer1M": round(provider_in, 4),
                "providerOutputUsdPer1M": round(provider_out, 4),
                "retailInputUsdPer1M": round(provider_in * markup, 4),
                "retailOutputUsdPer1M": round(provider_out * markup, 4),
                "onDemandInputUsdPer1M": round(provider_in * on_demand_markup, 4),
                "onDemandOutputUsdPer1M": round(provider_out * on_demand_markup, 4),
            }
        )
    return rows

File: functions/src/ai/usagePricing.ts
const DEFAULT_PROVIDER_USD_PER_1M: Record<string, [number, number]> = {
  "gpt-4.1-nano": [0.1, 0.4],
  "gpt-4o-mini": [0.15, 0.6],
  "grok-3-mini": [1.25, 2.5],
  "grok-4.1": [1.25, 2.5],
  "grok-4.3": [1.25, 2.5],
  "grok-cad-reasoning": [1.25, 2.5],
  "claude-opus-4-7": [5.0, 25.0],
  "claude-opus-4-8": [5.0, 25.0],
};

const MODEL_PRICING_ALIASES: Record<string, string> = {
  auto: "gpt-4o-mini",
  "grok-mini": "grok-3-mini",
  "grok-3-mini": "grok-3-mini",
  grok: "grok-4.1",
  "grok-4.1": "grok-4.1",
  "grok-4.3": "grok-4.3",
  "grok-4.20": "grok-cad-reasoning",
  "grok-4.20-reasoning": "grok-cad-reasoning",
  "grok-4.20-0309-reasoning": "grok-cad-reasoning",
  "grok-build-0.1": "grok-cad-reasoning",
  "gpt-4-1-nano": "gpt-4.1-nano",
  "gpt-4.1-nano": "gpt-4.1-nano",
  "gpt-4o": "gpt-4o-mini",
  "gpt-4o-mini": "gpt-4o-mini",
  "claude-opus-4-20250514": "claude-opus-4-7",
  "claude-opus-4-7": "claude-opus-4-7",
  "claude-opus-4-8": "claude-opus-4-8",
};

export interface ModelUsageRate {
  modelKey: string;
  providerInputUsdPer1M: number;
  providerOutputUsdPer1M: number;
  retailInputUsdPer1M: number;
  retailOutputUsdPer1M: number;
}

export function usageMarkupMultiplier(): number {
  const raw = process.env.FORMA_USAGE_MARKUP ?? "1.25";
  const value = Number.parseFloat(raw);
  return Number.isFinite(value) ? Math.max(value, 1) : 1.25;
}

export function onDemandUsageMarkupMultiplier(): number {
  const raw = process.env.FORMA_ON_DEMAND_USAGE_MARKUP ?? "1.65";
  const value = Number.parseFloat(raw);
  return Number.isFinite(value) ? Math.max(value, 1) : 1.65;
}

export function proUsageAllowanceUsd(): number {
  const raw = process.env.FORMA_PRO_USAGE_ALLOWANCE_USD ?? "30";
  const value = Number.parseFloat(raw);
  return Number.isFinite(value) ? Math.max(value, 0) : 30;
}

export function enterpriseUsageAllowancePerSeatUsd(): number {
  const raw = process.env.FORMA_ENTERPRISE_USAGE_ALLOWANCE_USD_PER_SEAT ?? "25";
  const value = Number.parseFloat(raw);
  return Number.isFinite(value) ? Math.max(value, 0) : 25;
}

export function enterpriseUsageAllowanceUsd(seatCount: number): number {
  const seats = Math.max(Number(seatCount) || 0, 1);
  return enterpriseUsageAllowancePerSeatUsd() * seats;
}

function parseRateEntry(value: unknown): [number, number] | null {
  if (typeof value === "number" && Number.isFinite(value)) {
    const per1M = value * 100;
    return [per1M, per1M];
  }
  if (value && typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const inp = obj.inputUsdPer1M ?? obj.input;
    const out = obj.outputUsdPer1M ?? obj.output;
    const inNum = Number(inp);
    const outNum = Number(out);
    if (Number.isFinite(inNum) && Number.isFinite(outNum)) return [inNum, outNum];
  }
  return null;
}

function providerTable(): Record<string, [number, number]> {
  const raw = (process.env.FORMA_MODEL_PROVIDER_RATES ?? "").trim();
  if (raw) {
    try {
      const parsed = JSON.parse(raw) as Record<string, unknown>;
      const out: Record<string, [number, number]> = {};
      for (const [key, value] of Object.entries(parsed)) {
        const rates = parseRateEntry(value);
        if (rates) out[key.trim().toLowerCase()] = rates;
      }
      if (Object.keys(out).length > 0) return out;
    } catch {
      /* ignore */
    }
  }

  const legacy = (process.env.FORMA_MODEL_PROVIDER_USD_PER_10K ?? "").trim();
  if (legacy) {
    try {
      const parsed = JSON.parse(legacy) as Record<string, unknown>;
      const out: Record<string, [number, number]> = {};
      for (const [key, value] of Object.entries(parsed)) {
        const rates = parseRateEntry(value);
        if (rates) out[key.trim().toLowerCase()] = rates;
      }
      if (Object.keys(out).length > 0) return out;
    } catch {
      /* ignore */
    }
  }

  return { ...DEFAULT_PROVIDER_USD_PER_1M };
}

export function normalizePricingModel(modelId: string): string {
  const mid = (modelId || "").trim().toLowerCase();
  if (!mid) return "gpt-4o-mini";
  if (MODEL_PRICING_ALIASES[mid]) return MODEL_PRICING_ALIASES[mid];
  if (mid.includes("nano") && mid.includes("gpt")) return "gpt-4.1-nano";
  if (mid.includes("gpt-4o-mini") || mid.startsWith("gpt-4o")) return "gpt-4o-mini";
  if (mid.includes("grok-3-mini") || mid === "grok-mini") return "grok-3-mini";
  if (mid.includes("build") || mid.includes("4.20") || mid.includes("reasoning")) {
    return "grok-cad-reasoning";
  }
  if (mid.includes("4.1") && mid.includes("grok")) return "grok-4.1";
  if (mid.includes("grok")) return "grok-4.3";
  if (mid.includes("opus") && (mid.includes("4-8") || mid.includes("48"))) return "claude-opus-4-8";
  if (mid.includes("claude") || mid.includes("opus")) return "claude-opus-4-7";
  return mid;
}

export function modelUsageRate(modelId: string): ModelUsageRate {
  const key = normalizePricingModel(modelId);
  const [inp, out] = providerTable()[key] ?? [1, 1];
  const markup = usageMarkupMultiplier();
  return {
    modelKey: key,
    providerInputUsdPer1M: inp,
    providerOutputUsdPer1M: out,
    retailInputUsdPer1M: inp * markup,
    retailOutputUsdPer1M: out * markup,
  };
}

export function usageCostUsd(
  modelId: string,
  inputTokens: number,
  outputTokens: number,
  retail = true,
  markupMultiplier?: number,
): number {
  const inp = Math.max(0, inputTokens);
  const out = Math.max(0, outputTokens);
  if (inp <= 0 && out <= 0) return 0;
  const rate = modelUsageRate(modelId);
  const multiplier = retail
    ? markupMultiplier ?? usageMarkupMultiplier()
    : 1;
  const providerCost =
    (inp / 1_000_000) * rate.providerInputUsdPer1M +
    (out / 1_000_000) * rate.providerOutputUsdPer1M;
  return providerCost * multiplier;
}

export function splitRetailCharge(
  providerCost: number,
  currentIncludedUsd: number,
  allowanceUsd: number,
  onDemandEnabled: boolean,
): [number, number] {
  if (providerCost <= 0) return [0, 0];
  const includedMarkup = usageMarkupMultiplier();
  const onDemandMarkup = onDemandUsageMarkupMultiplier();
  const remaining = Math.max(0, allowanceUsd - currentIncludedUsd);
  const fullAtIncluded = providerCost * includedMarkup;
  if (fullAtIncluded <= remaining + 1e-9) return [fullAtIncluded, 0];
  if (!onDemandEnabled) return [Math.min(fullAtIncluded, remaining), 0];
  if (remaining <= 1e-9) return [0, providerCost * onDemandMarkup];
  const includedRetail = remaining;
  const includedProvider = includedRetail / includedMarkup;
  const onDemandProvider = Math.max(0, providerCost - includedProvider);
  return [includedRetail, onDemandProvider * onDemandMarkup];
}
